DataScience/ImportantFeatures.py

```diff
@@ -0,0 +1,157 @@
+    if '\n' not in text:
+        print ('no features found in model output file: {0}'.format(fp))
+    else:
+        for line in text.splitlines():
+            data = line.split(':')
+            if data[0] in inv_hash:
+                features.append(inv_hash[data[0]])
+            else:
+                print ('missing hash value in inv_hash: {0}'.format(data[0]))
+    return features
+
+# read the invert hash file and return a dictionary that maps from hash value to feature.
+def get_feature_inv_hash(fp):
+    inv_hash = {}
+    text = open(fp).read().split('\n:0\n',1)[1].strip()
+    if '\n' not in text:
+        print ('no features found in invert has file: {0}.'.format(fp))
+    else:
+        for line in text.splitlines():
+            data = line.split(':')
+            if (len(data) == 3):
+                inv_hash[data[1]] = data[0]
+    return inv_hash
+    
+# return unique buckets of features from the feature funnel.
+# sample: input => [['c','b','a','d','e'],['b','c','a'],['a']] returns output => [['a'], ['b', 'c'], ['d', 'e']]
+def get_feature_buckets(features_funnel):
+    union_features = []
+    feature_buckets = []
+    for features in reversed(features_funnel):
+        unique_features = list(set(features) - set(union_features))
+        unique_features.sort()
+        feature_buckets.append(unique_features)
+        union_features.extend(unique_features)
+    return feature_buckets
+
+def get_important_features(log_file, ml_args, warmstart_model=None, min_num_features=5):
+    invHash_fp = log_file+'.invHash.txt'
+    readModel_fp_base = log_file+'.readModel.{0}.txt'
+ 
+    if ' --l1 ' in ml_args: 
+        temp = ml_args.split(' --l1 ',1)
+        ml_args = temp[0]
+        if ' ' in temp[1]:
+            temp = temp[1].split(' ',1)
+            ml_args += ' ' + temp[1]
+            l1 = float(temp[0])
+        else:
+            l1 = float(temp[1])
+    else:
+        l1 = 1e-7
+    
+    vw_base = 'vw ' + ml_args + ' --dsjson --data {0} --quiet'.format(log_file)
+    if warmstart_model:
+        vw_base += ' -i {0}'.format(warmstart_model)
+
+    print('\n=====================================')
+    print('Generating invert hash file to map the hash to feature names')
+
+    # Use a low l1 value to capture all features
+    vw_inv_hash_cmd = vw_base + ' --invert_hash {0} --l1 {1}'.format(invHash_fp, 1e-23)
+    print('command to get invert hash file: ' + vw_inv_hash_cmd)
+    os.system(vw_inv_hash_cmd)
+    inv_hash = get_feature_inv_hash(invHash_fp)
+
+    print('\n=====================================')
+    print('Testing a range of L1 regularization')
+    all_features_funnel = []
+    index = 0
+    max_run_count = 20
+    while True:
+        readModel_fp = readModel_fp_base.format(index)
+        vw_readable_model_cmd_base = vw_base + ' --readable_model {0}'.format(readModel_fp)
+        vw_readable_model_cmd = vw_readable_model_cmd_base + ' -c --l1 {0}'.format(index)
```

Comment:
Shouln't the l1 variable be used somewhere in this line?

Reply:
You are right. It is a bug.

---
In reply to: [279973367](https://github.com/Microsoft/mwt-ds/pull/132#discussion_r279973367) [](ancestors = 279973367)